### ISSUE TITLE
Allow invalid TLS certs for wss:// websocket_url

### ISSUE THREAD

**Issue by nordurljosahvida** [conversejs/converse.js#1583]
Not sure if this is entirely client-side, but would it be possible to allow self-signed certs for websocket_url? In my project I've allowed users to deploy two versions of converse so that the instance of converse accessed via .onion hostname [such as https://example.onion/xmpp/] uses `wss://example.onion/ws` as the websocket_url in order to conceal the clearnet domain of the server [[reference functions file in aenigma](https://github.com/openspace42/aenigma/blob/master/functions#L2741)]. I of course never endorse the use of invalid TLS certs, but in this case it's a must.

@jcbrand should make websocket_url default instead of bosh_service_url as it's soooo much better! ;]

Thanks!

this might help: https://github.com/mqttjs/MQTT.js/issues/288

**Comment by nordurljosahvida** [comment]
It occurred to me that of course when using converse via tor users are already accepting the self signed cert of `https://example.onion/xmpp/` where conversejs is hosted, so the "hidden" / "secondary" self-signed cert issue only happens when, as in our case, the websocket endpoint is on a different domain/subdomain than the pure top-level `example.onion` domain [such as xh01.example.onion in our case].

Therefore I decided to skip SSLH and hardcode the custom port [in our case 5443] in the websocket_url so that traffic went straight to ejabberd [host of the websocket], however firefox/Tor browser still warns about self-signed cert even if it's the same domain but a *different port*!

So in the end I setup nginx to reverse proxy websocket as so:

```
location /ws/ {
        proxy_pass https://127.0.0.1:5443/ws;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
    }
```

so that websocket_url can be exactly `example.onion/ws/`.

However one issue remains: the `example.onion/xmpp/` converse page can also be visited going to the hostname-specific version `ae01.example.onion/xmpp`, but then in that case the domain will be different since converse will always use `example.onion/ws` as the websocket_url.

**TL;DR: is there any way to make the websocket_url relative instead of absolute?**

such as: `websocket_url: /ws/` instead of `websocket_url: wss://example.onion/ws`?

Thanks!

**Comment by jcbrand** [comment]
Have you tried just making it relative? I think it might just work.

**Comment by nordurljosahvida** [comment]
I tried both `websocket_url: wss://ws/` and `websocket_url: /ws` but I think the former is invalid and Idon't think it applies the `wss` in the latter case.

**Comment by jcbrand** [comment]
Oh yeah of course, because it's a different protocol and therefore relative to the webpage's URL.

I think relative URLs will work for a BOSH connection though, if they're relative to the domain on which Converse is hosted.

I don't see how can make relative websocket URLs work, how do we know what they're relative to?

**Comment by nordurljosahvida** [comment]
@jcbrand not sure what you mean by "how do we know what they're relative to". I'm also not familiar with how they work really, so maybe that's why. Shouldn't the web browser make any relative link relative to the currently loaded page and therefore domain?

**Comment by jcbrand** [comment]
Websockets connections use a different transport protocol, so the browser can't handle relative websocket URLs.

**Comment by nordurljosahvida** [comment]
so this doesn't help right? https://stackoverflow.com/questions/10406930/how-to-construct-a-websocket-uri-relative-to-the-page-uri

**Comment by jcbrand** [comment]
Well, to be specific, that's not the browser handling a relative websocket URL, that's using JavaScript to construct a websocket URL by replacing the scheme of the current http URL.

This would be possible within Converse as well, but I'm not convinced that this should go into the core project.

You can write a plugin which overrides the `initConnection` method to do this, similarly to how the the `onConnected` method is overridden in this example code here: https://conversejs.org/docs/html/plugin_development.html#a-full-example-plugin

**Comment by nordurljosahvida** [comment]
ok thank you very much
